src/beamlime/applications/base.py

```diff
@@ -206,25 +149,17 @@ async def route(self, message: MessageProtocol) -> None:
         for result in results:
             self.message_pipe.put(result)
 
-    async def run(self) -> None:
+    async def run(
+        self,
+    ) -> AsyncGenerator[Optional[MessageProtocol], None]:
         """Message router daemon."""
-        data_monitor = self.data_pipe_monitor(
-            self.message_pipe,
-            timeout=0.1,
-            interval=0.1,
-            size_monitor_func=self.message_pipe.qsize,
-        )
-
-        while not self._break_routing_loop:
-            if await data_monitor():
-                message = self.message_pipe.get()
-                await self.route(message)
-                await asyncio.sleep(0)
-
-        self.debug("Breaking routing loop. Routing the rest of the message...")
-        while not self.message_pipe.empty():
-            await self.route(self.message_pipe.get())
-        self.debug("Routing the rest of the message done.")
+        while True:
```

Comment:
How does the loop get stopped?

Reply:
This is a generator so once it reaches yield, it's stopps there.

Reply:
It pauses, but doesn't stop and clean up completely. So does the caller simply drop its reference to the generator?

Reply:
Ah, now I understood...!

https://github.com/scipp/beamlime/blob/cfa63d08701e5f36f558ca4b7fe93ab1425e8e48/src/beamlime/applications/base.py#L222-L231

This method is used for creating a routine and it has a reference to ``self(Application instance)._break``.

So ``run_daemon`` coroutine is done, the generator should just stop...?

Reply:
Ok. It should be fine because `run` doesn't hold onto any resources that need to cleaned up.